Add wind power potential and wind profile indicators

huard opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #1458 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Adds two indices and their indicators: wind_profile and wind_power_potential. 

### Does this PR introduce a breaking change?
No


### Other information:
Ideally we`d add an indicator to compute air density, which is an optional input to the wind power potential.

laura-vanvliet commented on xclim/indices/_conversion.py:
```diff
@@ -2017,3 +2019,157 @@ def mean_radiant_temperature(
+    For non-standard air density (:math:`\rho`), the wind speed is scaled using
+    :math:`v_n = v \left( \frac{\rho}{\rho_0} \right)^{1/3}`.
+
+    Note that the temporal resolution of the wind speed time series has a significant influence on the results,
```
Perhaps this is not necessary, but you could also note that temporal resolution may affect aggregation in time

huard replied:
Changed to 

Note that the temporal resolution of the wind speed time series has a significant influence on the results,
    even when aggregated at longer time scales. Total annual power production will differ substantially if estimated
    from hourly or daily wind speeds, due to the non-linearity of the production factor.

Is this what you meant ?

laura-vanvliet commented on xclim/indices/_conversion.py:
```diff
@@ -2017,3 +2019,157 @@ def mean_radiant_temperature(
+    method : {"power_law"}
+        Method to use. Currently only "power_law" is implemented.
+    kwds : dict
+        Additional keyword arguments to pass to the method. For `power_law`, this is `alpha`, which takes a default
```
Could add some additional clarification, if desired: "For `power_law`, this is `alpha`, which takes a default value of 1/7, but is highly variable based on topography, surface cover and atmospheric stability. "

laura-vanvliet commented on tests/test_indices.py:
```diff
@@ -3626,3 +3626,24 @@ def test_late_frost_days_lat(self, tasmin_series):
+        p = xci.wind_power_potential(
+            sfcWind_series(a, units="m/s"), cut_in="4 m/s", rated="8 m/s"
+        )
+        np.testing.assert_allclose(p, [0, 0.5**3, 1, 0])
```
I assume 0.5**3 is the same as [ (v - ui) / (ur - ui) ]**3 for these specific values, but flagging just in case :)

huard replied:
Yeah it did, but not now with the change in the formula.

laura-vanvliet reviewed (changes requested):
Hi @huard, I think overall this looks great! There is one important change due to another typo in the Tobin (2018) paper, and some suggestions for the function descriptors.

huard commented on xclim/indices/_conversion.py:
```diff
@@ -2017,3 +2019,157 @@ def mean_radiant_temperature(
+    if v < cut_in:
+        return 0.0
+    if v < rated:
+        return ((v - cut_in) / (rated - cut_in)) ** 3
```
Got it, done.

huard commented on xclim/indices/_conversion.py:
```diff
@@ -2017,3 +2019,157 @@ def mean_radiant_temperature(
+
+        \begin{cases}
+        0,  &  v < u_i \\
+        (v - u_i)^3 / (u_r - u_i)^3,  & u_i ≤ v < u_r \\
```
Done

aulemahal commented on xclim/indices/_conversion.py:
```diff
@@ -5,7 +5,7 @@
 import numpy as np
 import xarray as xr
-from numba import float32, float64, vectorize  # noqa
+from numba import float32, float64, njit, vectorize  # noqa
```
```suggestion
from numba import float32, float64, vectorize  # noqa
```
I think this change is now superfluous ?

huard replied:
Indeed !

aulemahal reviewed (approved):
Looks good!

May be you could add real tests in `test_atmos` ?

laura-vanvliet commented:
Hi David, sorry for the long delay. I think the original wording was better. I was referring to, when summing up to create an annual power production estimate, they need to multiple by 3 if they use 3-hourly data, etc. But I'm not sure if that note is worth including.


From: David Huard ***@***.***>
Sent: Monday, September 18, 2023 9:02 AM
To: Ouranosinc/xclim ***@***.***>
Cc: laura-vanvliet ***@***.***>; Comment ***@***.***>
Subject: Re: [Ouranosinc/xclim] Add wind power potential and wind profile indicators (PR #1471)


@huard commented on this pull request.

________________________________

In xclim/indices/_conversion.py<https://github.com/Ouranosinc/xclim/pull/1471#discussion_r1328953386>:

> +    and rated speed, is constant between the rated and cut-out speed, and is zero for wind speeds above the cut-out

+    speed to avoid damage to the turbine :cite:p:`tobin_2018`:

+

+    .. math::

+

+        \begin{cases}

+        0,  &  v < u_i \\

+        (v - u_i)^3 / (u_r - u_i)^3,  & u_i ≤ v < u_r \\

+        1, & u_r ≤ v < u_o \\

+        0, & v ≥ u_o

+        \end{cases}

+

+    For non-standard air density (:math:`\rho`), the wind speed is scaled using

+    :math:`v_n = v \left( \frac{\rho}{\rho_0} \right)^{1/3}`.

+

+    Note that the temporal resolution of the wind speed time series has a significant influence on the results,

Changed to

Note that the temporal resolution of the wind speed time series has a significant influence on the results,
even when aggregated at longer time scales. Total annual power production will differ substantially if estimated
from hourly or daily wind speeds, due to the non-linearity of the production factor.

Is this what you meant ?

-
Reply to this email directly, view it on GitHub<https://github.com/Ouranosinc/xclim/pull/1471#discussion_r1328953386>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASLWXWBA5XVYBFH7ANHCAB3X3BV6PANCNFSM6AAAAAA4Q3FM7U>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>

huard commented:
Ok, changed the wording of the docstring to simplify, and added a note explaining how to aggregate instantaneous power potential at lower frequencies. Tests have also been improved in the process, including one demoing the conversion to daily power production in MWh from 3-hourly values. 

Also fixed a bug in to_agg_units, but I note there are no explicit tests for this function.
